subsystem/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): L. Kuklinek
+
+//! General framework for working with subsystems
+//!
+//! The [Manager] type handles a collection of subsystems. The framework also takes care of
+//! inter-subsystem calls and clean shutdown. Subsystems communicate using [Handle]s.
+//!
+//! ## Calls
+//!
+//! Calls are dispatched by sending a closure over a channel to the subsystem. The subsystem then
+//! sends the result back using a oneshot channel. The channel is awaited to emulate synchronous
+//! calls.
+//!
+//! ## Shutdown sequence
+//!
+//! The shutdown proceeds in three phases:
+//!
+//! 1. As soon as any subsystem terminates, the main task is notified.
+//! 2. The main task broadcasts the shutdown request to all subsystems. The subsystems react to the
+//!    request by shutting themselves down.
+//! 3. The main task waits for all subsystems to terminate.
+
+use core::future::Future;
+use core::time::Duration;
+use futures::future::BoxFuture;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio::task;
+
+use logging::log;
+
+/// Defines hooks into a subsystem lifecycle.
+#[async_trait::async_trait]
+pub trait Subsystem: Sized {
+    /// Custom shutdown procedure.
+    async fn shutdown(self) {}
+}
+
+/// Manager configuration options.
+pub struct ManagerConfig {
```

Comment:
May I suggest moving things around and making this design more modular? Maybe the manager can be moved to a submodule. The file is already 500 lines.

Reply:
Yep very good suggestion. Will do. Although moving files around while a PR is open makes reviewing more difficult, so will only go ahead if you and others are fine with doing it now.

Reply:
Actually, I can use my time zone advantage and do it today in the evening, when others are unlikely to be looking at it.